doc/whatsnew/2.13.rst

```diff
@@ -37,6 +37,12 @@ New checkers
 
   Closes #5460
 
+* ``modified-iterating-list``: Emitted when items are added to or removed from a list being iterated through.
+* ``modified-iterating-dict``: Emitted when items are added to or removed from a dict being iterated through.
+* ``modified-iterating-set``: Emitted when items are added to or removed from a set being iterated through.
```

Comment:
@DanielNoord  - The message was at the `remove/add` before but I decided to change it to the `for` line. The reason is that the actual thing you need to fix is the `for` line. I agree with you though that the user wouldn't know for sure why this message was raised for long `for` loops, so I suggest modifying the message so that the `add/remove` line/line number is shown there. also we can get rid of the "only one message per `for` loop" logic and raise more than 1 message if necessary since after adding the 'problematic' `add/remove` line makes the messages different.

* for pop/update - we can open an issue for extending this checker after this is merged.

Reply:
well removing the `add/remove` changes the logic of your code and you don't want to suggest the user to do that in your message. iterating through a copy of the `list` in the `for` loop keeps the logic the same.
Let's play with an example, say you have a list of numbers and for each nonzero-1-digit number in the list you want to modify it to a 2-digit number and have copies of the new number times the value of the original, so you'd write something like this:

```
lst = [1, 2, 3]
for item in lst:
    if item//10 == 0:
        lst.remove(item)
        for i in range(item):
            lst.append((item * 10) + item)
```
You expect:

`lst = [11, 22, 22, 33, 33, 33]`

but you get:
`lst = [2, 11, 33, 33, 33]`

 The code has unexpected behavior because of our bug. you need the `add/remove` lines because this is what you want to do.. iterating through the copy of the list gives the expected value in `lst`

Reply:
I agree that his is the bug we're trying to solve here. However, I still don't think we should raise on the `for` loop:
1) The actual offending line is the `add/remove`, I believe we should always raise on what is wrong instead of where to solve.
2) If we raise on the `add/remove` line it's trivial to find the `for` loop (indentation), the other way around is more difficult. There are situations where providing the `lineno` doesn't really help in finding the offending line (see `Zen` mode in VS Code)
3) _It is also more in line with the message name, the list/dict/set gets modified on the `add/remove` line._

I can't come up with an example for this, but I think there are other examples within `pylint` where the solution to a message is not necessarily on the line where the message is raised. It also helps python beginners why something is raised, which is probably also the target for this message. (More experienced users aren't likely to do this).

Reply:
I agree with Daniel here, pylint always point to the problem, not to the solution. It sometime suggest a solution in its message but it would be counterintuitive to start doing this.

Reply:
Okay I would change to the `add/remove` line.. then we could have more than 1 message for the same `for` loop.
I still think the solution suggestion is helpful and should be included in the message.

A small suggestion to improve the message, adding the word `for`:

Iterated list '%s' is being modified inside **for** loop body, consider iterating through a copy of it instead.

Reply:
Very nice message 👍